tests/test_general.py

```diff
@@ -27,899 +26,941 @@
+
+
+def test_vector_fitting_guess():
+    """
+    Tests fitting to a 3 component vector valued function, with guesses.
+    """
+    a, b, c = parameters('a, b, c')
+    a.value = 10
+    b.value = 100
+    a_i, b_i, c_i = variables('a_i, b_i, c_i')
+
+    model = {a_i: a, b_i: b, c_i: c}
+
+    xdata = np.array([
+        [10.1, 9., 10.5, 11.2, 9.5, 9.6, 10.],
+        [102.1, 101., 100.4, 100.8, 99.2, 100., 100.8],
+        [71.6, 73.2, 69.5, 70.2, 70.8, 70.6, 70.1],
+    ])
+
+    fit = Fit(
+        model=model,
+        a_i=xdata[0],
+        b_i=xdata[1],
+        c_i=xdata[2],
+        minimizer=MINPACK
+    )
+    fit_result = fit.execute()
+
+    assert fit_result.value(a) == pytest.approx(np.mean(xdata[0]), 1e-4)
+    assert fit_result.value(b) == pytest.approx(np.mean(xdata[1]), 1e-4)
+    assert fit_result.value(c) == pytest.approx(np.mean(xdata[2]), 1e-4)
+
+
+def test_fitting():
+    """
+    Tests fitting with NumericalLeastSquares. Makes sure that the resulting
+    objects and values are of the right type, and that the fit_result does
+    not have unexpected members.
+    """
+    xdata = np.linspace(1, 10, 10)
+    ydata = 3*xdata**2
+
+    a = Parameter()  # 3.1, min=2.5, max=3.5
+    b = Parameter()
+    x = Variable()
+    new = a*x**b
+
+    fit = Fit(new, xdata, ydata, minimizer=MINPACK)
+
+    fit_result = fit.execute()
+    assert isinstance(fit_result, FitResults)
+    assert fit_result.value(a) == pytest.approx(3.0)
+    assert fit_result.value(b) == pytest.approx(2.0)
+
+    assert isinstance(fit_result.stdev(a), float)
+    assert isinstance(fit_result.stdev(b), float)
+
+    assert isinstance(fit_result.r_squared, float)
+    assert fit_result.r_squared == 1.0  # by definition since there's no fuzzyness
+
+
+def test_grid_fitting():
+    """
+    Tests fitting a scalar function with 2 independent variables.
+    """
+    xdata = np.arange(-5, 5, 1)
+    ydata = np.arange(5, 15, 1)
+    xx, yy = np.meshgrid(xdata, ydata, sparse=False)
+
+    zdata = (2.5*xx**2 + 3.0*yy**2)
+
+    a = Parameter(value=2.5, max=2.75)
+    b = Parameter(value=3.0, min=2.75)
+    x = Variable('x')
+    y = Variable('y')
+    z = Variable('z')
+    new = {z: a*x**2 + b*y**2}
+
+    fit = Fit(new, x=xx, y=yy, z=zdata)
+    results = fit.execute()
+
+    assert isinstance(fit.minimizer, LBFGSB)
+
+    assert results.value(a) == pytest.approx(2.5)
+    assert results.value(b) == pytest.approx(3.)
+
+# TODO: Should be 3 tests?
+
+
+def test_model_callable():
+    """
+    Tests if Model objects are callable in the way expected. Calling a
+    model should evaluate it's expression(s) with the given values. The
+    return value is a namedtuple.
+
+    The signature should also work so inspection is saved.
+    """
+    a, b = parameters('a, b')
+    x, y = variables('x, y')
+    new = a*x**2 + b*y**2
+    model = Model(new)
+    ans = model(3, 3, 2, 2)
+    assert isinstance(ans, tuple)
+    z, = ans
+
+    assert z == 36
```

Comment:
I would replace the 36 with `2*3**2 + 2*3**2`. But it can be left for later.
Also note that this test is not that good, since x and y could be swapped, and we'd never see it.

Reply:
Allright. I'll skip that first